### PR TITLE
[screenjunkies] Add new extractor

### PR DESCRIPTION
**Pull request by TingPing**

### CODE REVIEW

File: youtube_dl/extractor/screenjunkies.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import re
+import json
+
+from .common import (
+    InfoExtractor,
+    ExtractorError
+)
+
+class ScreenJunkiesIE(InfoExtractor):
+    _VALID_URL = r'http://www.screenjunkies.com/video/(?P<id>.*)'
```
Comment (dstftw) [inline youtube_dl/extractor/screenjunkies.py]: `id` must match at least one character.

File: youtube_dl/extractor/screenjunkies.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            'title': "Honest Trailers - 'The Dark Knight'",
+        },
+    }, {
+        'url': 'http://www.screenjunkies.com/video/knocking-dead-ep-1-the-show-so-far-3003285',
+        'only_matching': True,
+    }]
+
+    def _real_extract(self, url):
+        video_link = self._match_id(url)
+        video_id = self._search_regex(r'.*-(\d+)', url, 'id', default=None, fatal=False)
```
Comment (dstftw) [inline youtube_dl/extractor/screenjunkies.py]: `id` should be extracted from `_VALID_URL` with `self._match_id`.

File: youtube_dl/extractor/screenjunkies.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+        if not video_id: # Older urls didn't have the id in them, but we can grab it manually
+            webpage = self._download_webpage(url, video_link)
+            video_id = self._search_regex(r'src="/embed/(\d+)"', webpage, 'video id')
+
+        webpage = self._download_webpage('http://www.screenjunkies.com/embed/%s' %video_id, video_id)
+        video_vars_str = self._html_search_regex(r'embedVars = (\{.+\})\s*</script>', webpage, 'video variables', flags=re.S)
+        video_vars = json.loads(video_vars_str)
```
Comment (dstftw) [inline youtube_dl/extractor/screenjunkies.py]: Use `_parse_json` instead.

File: youtube_dl/extractor/screenjunkies.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+        # TODO: Figure out required cookies
+        if video_vars['subscriptionLevel'] > 0:
+            raise ExtractorError('This video requires ScreenJunkiesPlus')
```
Comment (dstftw) [inline youtube_dl/extractor/screenjunkies.py]: `expected=True`.

File: youtube_dl/extractor/screenjunkies.py (hunk 2 above)
Comment (TingPing) [inline youtube_dl/extractor/screenjunkies.py]: Not all URLs have the id included though.

File: youtube_dl/extractor/screenjunkies.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+                'width': f['width'],
+                'height': f['height'],
+                'format': 'mp4',
+            })
+        formats.sort(key=lambda d: d['tbr'])
+
+        return {
+            'id': video_id,
+            'title': video_vars['contentName'],
+            'thumbnail': video_vars['thumbUri'],
```
Comment (dstftw) [inline youtube_dl/extractor/screenjunkies.py]: Extraction should be tolerate to missing fields. Everything apart from title and formats considered optional and should not break the extraction if missing.

File: youtube_dl/extractor/screenjunkies.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            'duration': video_vars['videoLengthInSeconds'],
```
Comment (dstftw) [inline youtube_dl/extractor/screenjunkies.py]: `int_or_none`.

File: youtube_dl/extractor/screenjunkies.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+            raise ExtractorError('This video requires ScreenJunkiesPlus')
+
+        formats = []
+        for f in video_vars['media']:
+            if not f['mediaPurpose'] == 'play':
+                continue
+
+            formats.append({
+                'url': f['uri'],
+                'tbr': f['bitRate'],
```
Comment (dstftw) [inline youtube_dl/extractor/screenjunkies.py]: `int_or_none`.

File: youtube_dl/extractor/screenjunkies.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                'width': f['width'],
```
Comment (dstftw) [inline youtube_dl/extractor/screenjunkies.py]: `int_or_none`.

File: youtube_dl/extractor/screenjunkies.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+                'height': f['height'],
```
Comment (dstftw) [inline youtube_dl/extractor/screenjunkies.py]: `int_or_none`.

File: youtube_dl/extractor/screenjunkies.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+            })
+        formats.sort(key=lambda d: d['tbr'])
+
+        return {
+            'id': video_id,
+            'title': video_vars['contentName'],
+            'thumbnail': video_vars['thumbUri'],
+            'duration': video_vars['videoLengthInSeconds'],
+            'formats': formats,
+            'tags': video_vars['tags'],
```
Comment (dstftw) [inline youtube_dl/extractor/screenjunkies.py]: Must be list, not a string.

File: youtube_dl/extractor/screenjunkies.py (hunk 2 above)
Comment (dstftw) [inline youtube_dl/extractor/screenjunkies.py]: >when present.

**Comment by dstftw** [comment]
Cherry-picked in b9e8d7140aa8584f8516c3575c20b21172278fef, thanks.
Do you want to be credited as one of the authors of youtube-dl? If so under what name?

**Comment by TingPing** [comment]
> Do you want to be credited as one of the authors of youtube-dl? If so under what name?

sure: Patrick Griffis

**Comment by TingPing** [comment]
@dstftw Also thanks a ton for getting the Plus only videos working!

**Comment by dstftw** [comment]
Done in fe80df30804736db5e644ad15ac09149c57ddf33.

**Pull request closed without merging**